pkg/plugin/conntrack/_cprog/conntrack.h

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#pragma once
+
+#include "vmlinux.h"
+#include "bpf_helpers.h"
+
+// Helper functions to get the current time
+// Ref: https://github.com/cilium/cilium/blob/6186d579ed60f334c7a4daaf81060797b02cc6bd/bpf/lib/time.h
+#define NSEC_PER_SEC	(1000ULL * 1000ULL * 1000UL)
+#define bpf_ktime_get_sec()	\
+	({ __u64 __x = bpf_ktime_get_boot_ns() / NSEC_PER_SEC; __x; })
+# define bpf_mono_now()		bpf_ktime_get_sec()
+
+// These units are in seconds
+
+// Define how long a TCP connection should be kept in the table
+#define CT_CONNECTION_LIFETIME_TCP 360 
+// Define how long a non-TCP connection should be kept in the table
+#define CT_CONNECTION_LIFETIME_NONTCP 60
+// Define how long a TCP connection should be kept alive after receiving the first SYN
+#define CT_SYN_TIMEOUT 60
+// Define the interval at which a packet should be sent to the userspace
+#define CT_REPORT_INTERVAL 30
+// Define the maximum number of connections that can be stored in the conntrack table
+#define CT_MAP_SIZE 262144
+
+enum tcp_flags {
+    TCP_FIN = 0x01,
+    TCP_SYN = 0x02,
+    TCP_RST = 0x04,
+    TCP_PSH = 0x08,
+    TCP_ACK = 0x10,
+    TCP_URG = 0x20,
+    TCP_ECE = 0x40,
+    TCP_CWR = 0x80,
+    TCP_NS  = 0x100
+};
+
+enum ct_packet_dir {
```

Comment:
For all these enums, I suggest coalescing into bit fields.
(optionally for extra compactness could consider stuffing them all in one large type.

Reply:
not sure what you mean by this, can you add a suggestion?

Reply:
Convert all these enums with few values into a more compact representation like you do for some other types - these can and should all be bit masks :)

Reply:
`ct_packet_dir` `ct_traffic_dir` and `obs_point` can have all of their values represented with 1+2+2=5 bits, but you're using 3*32=66 bits instead, so over 12x redundancy

Reply:
The first-level optimization is to use a single byte (8 bit) variable for each enum when passing these as parameters - that might be equally readable as separate enums.

The second-level optimization I meant, is that all these 3 enums can be at once represented with 2 bytes: total of 9 values can occupy a single bit each in a 16bit half-word (uint16_t).

Theird-level optimization, is that if that variables of these enums types can never have more than one value at once (unlike a packet having many tcp flags set, direction is only one per packet etc..) then you don't even need a 1 bit per value, but they can be coded as combinations: eg obs_point would only need 2 bits, rather than 4.

Lastly, all these types could be stored in a single vector, just as a CPU uses 64bit word and interprets sub-sections of them as instruction op code, parameters, registers etc. I'd do this with some discernment keeping in mind which types are semantically related tho. Happy to chat offline line if it 's still unclear :)

Reply:
thanks for the explanation! I think I'll go with the first level, which is to make everything a mask, which alleviate the needs for enums as well as still keeping the code readable 🙂